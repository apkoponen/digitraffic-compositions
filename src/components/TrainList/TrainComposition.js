import React, { Component } from 'react';
import Skeleton from '../Skeleton/Skeleton';
import './TrainComposition.css';

class TrainComposition extends Component {
  formatLocomotive = locomotives => {
    return locomotives
      .map(locomotive => `${locomotive.locomotiveType} ${locomotive.powerType}`)
      .join(', ');
  };

  formatWagons = (wagons, length) => {
    const formattedWagons = wagons
      .filter(wagon => wagon.wagonType)
      .map(wagon => wagon.wagonType)
      .join(', ');
    return `${formattedWagons} (${length} meters)`;
  };

  render() {
    const { composition, skeleton } = this.props;
    let locomotive, wagons, maxSpeed, originStation, destinationStation;
    if (skeleton) {
      locomotive = <Skeleton width="50" />;
      wagons = <Skeleton width="100" />;
      maxSpeed = <Skeleton width="50" />;
      originStation = <Skeleton width="25" />;
      destinationStation = <Skeleton width="25" />;
    } else {
      locomotive = this.formatLocomotive(composition.locomotives);
      wagons = this.formatWagons(composition.wagons, composition.totalLength);
      maxSpeed = composition.maximumSpeed + ' km/h';
      originStation = composition.beginTimeTableRow.stationShortCode;
      destinationStation = composition.endTimeTableRow.stationShortCode;
    }

    return (
      <div className="TrainComposition">
        <div className="TrainComposition__content">
          <p>
            <strong>Locomotive:</strong> {locomotive}
          </p>
          <p>
            <strong>Wagons:</strong> {wagons}
          </p>
          <p>
            <strong>Max speed:</strong> {maxSpeed}
          </p>
        </div>
        <div className="TrainComposition__timeline">
          <div className="TrainComposition__origin-station">
            {originStation}
          </div>
          <div className="TrainComposition__destination-station">
            {destinationStation}
          </div>
        </div>
      </div>
    );
  }
}

export default TrainComposition;
