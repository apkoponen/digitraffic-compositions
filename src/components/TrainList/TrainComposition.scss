@import 'styles/_colors';

$timeline-color: #666;

.TrainComposition {
  display: flex;
  align-items: stretch;

  p {
    margin-top: 0;
    margin-bottom: 5px;
  }
}

.TrainComposition__content {
  padding: 10px;
  font-size: 14px;
  flex: 1;
  background: #f6f6f6;
}

.TrainComposition__timeline {
  z-index: 2;
  flex: 0 0 30px;
  margin-left: 3px;
  border-left: 2px solid $timeline-color;
  position: relative;

  &:before, &:after {
    content: '';
    width: 8px;
    height: 8px;
    border-radius: 50%;
    position: absolute;
    left: -5px;
    background: $timeline-color;
    border: 2px solid #fff;
  }

  &:before {
    top: -2px;
  }

  &:after {
    bottom: -2px;
  }
}

.TrainComposition__origin-station, .TrainComposition__destination-station {
  position: absolute;
  left: 3px;
  color: #333;
  font-size: 12px;
}

.TrainComposition__origin-station {
  top: 0;
}

.TrainComposition__destination-station {
  bottom: 0;
}