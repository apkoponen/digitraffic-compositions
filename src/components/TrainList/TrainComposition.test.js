import React from 'react';
import TrainComposition from './TrainComposition';
import renderer from 'react-test-renderer';
import trains from '../../test-data/trains';

it('renders correctly', () => {
  const tree = renderer
    .create(<TrainComposition composition={trains[0].compositions[0]} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders skeleton correctly', () => {
  const tree = renderer.create(<TrainComposition skeleton />).toJSON();
  expect(tree).toMatchSnapshot();
});
