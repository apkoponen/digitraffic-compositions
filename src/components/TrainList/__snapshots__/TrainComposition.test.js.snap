// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="TrainComposition"
>
  <div
    className="TrainComposition__content"
  >
    <p>
      <strong>
        Locomotive:
      </strong>
       
      Sr2 Electric
    </p>
    <p>
      <strong>
        Wagons:
      </strong>
       
      Edfs, Edb, Edo (99 meters)
    </p>
    <p>
      <strong>
        Max speed:
      </strong>
       
      160 km/h
    </p>
  </div>
  <div
    className="TrainComposition__timeline"
  >
    <div
      className="TrainComposition__origin-station"
    >
      HKI
    </div>
    <div
      className="TrainComposition__destination-station"
    >
      TKU
    </div>
  </div>
</div>
`;

exports[`renders skeleton correctly 1`] = `
<div
  className="TrainComposition"
>
  <div
    className="TrainComposition__content"
  >
    <p>
      <strong>
        Locomotive:
      </strong>
       
      <span
        className="Skeleton"
        style={
          Object {
            "width": "50px",
          }
        }
      />
    </p>
    <p>
      <strong>
        Wagons:
      </strong>
       
      <span
        className="Skeleton"
        style={
          Object {
            "width": "100px",
          }
        }
      />
    </p>
    <p>
      <strong>
        Max speed:
      </strong>
       
      <span
        className="Skeleton"
        style={
          Object {
            "width": "50px",
          }
        }
      />
    </p>
  </div>
  <div
    className="TrainComposition__timeline"
  >
    <div
      className="TrainComposition__origin-station"
    >
      <span
        className="Skeleton"
        style={
          Object {
            "width": "25px",
          }
        }
      />
    </div>
    <div
      className="TrainComposition__destination-station"
    >
      <span
        className="Skeleton"
        style={
          Object {
            "width": "25px",
          }
        }
      />
    </div>
  </div>
</div>
`;
